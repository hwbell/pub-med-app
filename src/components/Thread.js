import React from 'react';
import '../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import Comment from './Comment';
import { Button, Fade, Collapse } from 'reactstrap';
import CommentForm from './CommentForm';
import AlertModal from './AlertModal';

// enter / exit transitions
import { CSSTransitionGroup } from 'react-transition-group' // ES6

// functions
import { deleteThread } from '../tools/serverFunctions';
import { extractStringDate } from '../tools/objectFunctions';

// ******************************************************************************
class Thread extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showCommentForm: false,
      showDeleteWarning: false,
      showContent: false
    }

    this.renderThread = this.renderThread.bind(this);
    this.toggleContent = this.toggleContent.bind(this);
    this.renderComments = this.renderComments.bind(this);
    this.toggleCommentForm = this.toggleCommentForm.bind(this);
    this.renderCommentForm = this.renderCommentForm.bind(this);
    this.submitComment = this.submitComment.bind(this);
    this.toggleDeleteWarning = this.toggleDeleteWarning.bind(this);
    this.handleDeleteThread = this.handleDeleteThread.bind(this);
  }

  toggleContent() {
    this.setState({
      showContent: !this.state.showContent
    }, () => {
      console.log(this.state.showContent)
    })
  }

  toggleCommentForm() {
    this.setState({
      showCommentForm: !this.state.showCommentForm
    })
  }

  toggleDeleteWarning() {
    console.log('toggling delete warning!')
    this.setState({
      showDeleteWarning: !this.state.showDeleteWarning
    })
  }

  handleDeleteThread() {

    // for server requests
    const token = localStorage.getItem('token') !== "undefined" ? JSON.parse(localStorage.getItem('token')) : null;
    const headers = {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    }

    return deleteThread(headers, this.props.thread)
      .then((response) => {
        console.log(response)

        // once we have posted, we will refresh in App
        this.props.deleteThread(this.props.thread);

        if (this.state.showDeleteWarning) {
          this.toggleDeleteWarning();
        }

        // Fail => show the fail message / reason

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  renderCommentForm() {
    return (
      <Collapse isOpen={this.state.showCommentForm}>
        <CommentForm toggle={this.toggleCommentForm} confirm={this.submitComment} />
      </Collapse>
    )
  }

  submitComment(comment) {
    // console.log(comment)
    // console.log(`submitting ${comment.user.name}'s comment to ${this.props.thread.name} thread:
    // ${comment.text}`)

    comment._id = this.props.thread._id;
    comment.owner = this.props.thread.owner;
    this.props.handleSubmitThread(comment);
  }

  renderThread() {

    let thread = JSON.parse(JSON.stringify(this.props.thread));

    let createdAt = new Date(thread.createdAt)
    let time = extractStringDate(createdAt);

    return (
      <div style={styles.threadContainer} className="thread">

        <div className="space-all-row">
          <p className="thread-title" style={styles.title}>{thread.name}</p>
          {/* these will show if the user owns the thread */}
          {this.props.allowEdit &&
            <div style={styles.buttonHolder}>
              <i className="fas fa-edit article-button"
                onClick={() => this.props.handleEdit(thread)}></i>

              <i className="fas fa-trash-alt article-button"
                onClick={this.toggleDeleteWarning}></i>

            </div>}
        </div>

        <p className="thread-text">
          <i className=" fas fa-user"></i>{`  ${thread.user}`}
        </p>

        <p className="time" >{time}</p>
        <p className="time" >
          {`@ article(s): ${thread.article}`}
        </p>

        {thread.paragraph && <p className="thread-text" >{thread.paragraph}</p>}

        {/* the content expander button */}
        <div className="left-all-row">
          <p className="thread-text">{`${thread.commentsCount} comments`}</p>

          <Button
            style={styles.expandButton}
            color="link" size="md"
            onClick={this.toggleContent}>
            <i className={this.state.showContent ? "fas fa-angle-double-up" : "fas fa-angle-double-down"}></i>
          </Button>
        </div>

        <Collapse style={{ width: '100%' }} isOpen={this.state.showContent}>

          <CSSTransitionGroup
            style={styles.transitionGroup}
            transitionName="drop"
            transitionEnterTimeout={500}
            transitionLeaveTimeout={300}>

            {thread.comments &&
              this.renderComments(thread)}

            {/* the boolean for this is inside the function, since we are using a Collapse element */}
            {this.renderCommentForm()}

            {/* this uses the same boolean, flipped */}
            {this.props.user &&
              <Fade style={styles.button} in={!this.state.showCommentForm}>
                <Button disabled={this.state.showCommentForm} className="view article-button" size="sm"
                  onClick={this.toggleCommentForm}>comment</Button>
              </Fade>}

          </CSSTransitionGroup>
        </Collapse>
      </div>
    )
  }

  renderComments(thread) {
    return thread.comments.map((comment, i) => {

      return (
        <Comment key={i} comment={comment} />
      )

    })
  }

  render() {

    return (
      <div className="outline">

        <AlertModal
          message={`Are you sure you want to delete this thread?`}
          isVisible={this.state.showDeleteWarning}
          confirm={this.handleDeleteThread}
          confirming={true}
          toggle={this.toggleDeleteWarning}
        />

        {this.renderThread(this.props.thread)}

      </div>
    );
  }
}

const styles = {
  threadContainer: {
    width: '100%',
    // padding: '10px',
    // paddingTop: '20px'
  },
  title: {
    width: '75%'
  },
  button: {
    alignSelf: 'flex-end',
  },
  expandButton: {
    margin: '4px'
  },
  transitionGroup: {
    paddingTop: '20px',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start'
  },
  buttonHolder: {
    // position: 'absolute',
    // top: '10px',
    // right: '10px',
    alignSelf: 'flex-start',
    display: 'flex',
    flexDirection: 'row'
  },
  hr: {
    backgroundColor: 'rgba(245,245,245,0.1)',
    margin: '4px',
    padding: 0
  }
}

export default Thread;
