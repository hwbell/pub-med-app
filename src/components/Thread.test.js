import React from 'react';

import Thread from './Thread';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// stubs
const toggleThreadFormStub = jest.fn();

describe('Thread', () => {

  let someProps;
  let thread;
  let wrapper;

  beforeEach(() => {
    thread = {
      name: 'Sourcing of PMID: 013091283',
      article: '013091283',
      user: 'The Creator',
      paragraph: 'I found the source material for the basis of this study to be lacking. Does anyone else agree?',
      comments: [
        {
          text: 'This article seems to lack source material for the conclusions drawn in the discussion.',
          user: 'Jeff'
        },
        {
          text: 'Does anyone else find figure 5 really confusing?',
          user: 'mark'
        }
      ],
      commentsCount: 2
    }
    someProps = {
      user: {
        name: 'mark',
        _id: '9898y23e23ee20389u'
      },
      thread,
      toggleThreadForm: toggleThreadFormStub
    };

    wrapper = shallow(<Thread {...someProps}/>);
  })

  it('renders without crashing', async () => {
    wrapper = shallow(<Thread {...someProps}/>);
  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<Thread  {...someProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct elements', () => {
    expect(wrapper.find('.outline').length).toBe(1)

    expect(wrapper.find('.thread-title').length).toBe(1)
    expect(wrapper.find('.thread-text').length).toBe(3)
    expect(wrapper.find('.view').length).toBe(1)
    expect(wrapper.find('Fade').length).toBe(1)
    expect(wrapper.find('AlertModal').length).toBe(1);

    // these three cover information that we would always have - the name, article, and user
    expect(wrapper.find('.thread-title').text()).toBe(thread.name);

    expect(wrapper.find('.thread-text').at(0).text()).toBe(`  ${thread.user}`);

    // this one we may or may not have
    let paragraph = thread.paragraph.slice(0,50);
    expect(wrapper.find('.thread-text').at(1).text()).toBe(paragraph + ` ...`);

  })

  it('should not display the comment button if there is no user in props', () => {
    wrapper.setProps({
      user: null
    });
    wrapper.update();

    expect(wrapper.find('.view').length).toBe(0);
  })

  it('should toggle the comment form', () => {
    expect(wrapper.state().showCommentForm).toBe(false);

    wrapper.instance().toggleCommentForm();
    wrapper.update();

    expect(wrapper.state().showCommentForm).toBe(true);
    
    wrapper.instance().toggleCommentForm();
    wrapper.update();

    expect(wrapper.state().showCommentForm).toBe(false);
  })

  it('should change hide the comment button when it is clicked', () => {
    wrapper.find('.view').simulate('click');
    wrapper.update();

    expect(wrapper.state().showCommentForm).toBe(true);
    expect(wrapper.find('Fade').props().in).toBe(false);
  })

  it('should change / hide the Collapse holding CommentForm when the comment button it is clicked', () => {

    expect(wrapper.find('Collapse').props().isOpen).toBe(false);

    wrapper.find('.view').simulate('click');
    wrapper.update();

    expect(wrapper.state().showCommentForm).toBe(true);
    expect(wrapper.find('Collapse').props().isOpen).toBe(true);
  })

  it('should render delete + edit buttons if its a user-owned thread', () => {
    expect(wrapper.find('.fa-edit').length).toBe(0);
    expect(wrapper.find('.fa-trash-alt').length).toBe(0);

    wrapper.setProps({
      allowEdit: true
    });

    wrapper.update();

    expect(wrapper.find('.fa-edit').length).toBe(1);
    expect(wrapper.find('.fa-trash-alt').length).toBe(1);
  })

  it('should toggle the showDeleteWarning boolean when the trash icon is clicked', () => {
    // get the buttons in there
    wrapper.setProps({
      allowEdit: true
    });
    wrapper.update();

    expect(wrapper.state().showDeleteWarning).toBe(false);

    wrapper.find('.fa-trash-alt').simulate('click');
    wrapper.update();

    expect(wrapper.state().showDeleteWarning).toBe(true);
  })

  it('should show how many comments a thread has', () => {
    expect(wrapper.find('Comment').length).toBe(2);

    expect(wrapper.find('.thread-text').at(2).text()).toBe(`${thread.commentsCount} comments`)
  })

  it('should toggle the ThreadForm when the edit button is clicked', () => {
    jest.clearAllMocks();

    // get the buttons in there
    wrapper.setProps({
      allowEdit: true
    });
    wrapper.update();
    
    wrapper.find('.fa-edit').simulate('click');

    expect(toggleThreadFormStub.mock.calls.length).toBe(1)
  
  })


})