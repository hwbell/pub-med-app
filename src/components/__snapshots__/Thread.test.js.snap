// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thread renders correctly 1`] = `
<div
  className="outline"
>
  <div
    style={
      Object {
        "alignItems": "flex-start",
        "alignSelf": "flex-start",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "flex-start",
        "padding": "10px",
        "width": "100%",
      }
    }
  >
    <p
      className="thread-title"
    >
      Sourcing of PMID: 013091283
    </p>
    <p
      className="thread-text"
    >
      <i
        className=" far fa-user"
      />
        The Creator
    </p>
    <p
      className="thread-text"
    >
      I found the source material for the basis of this  ...
    </p>
    <span
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "flex-start",
          "padding": "20px",
        }
      }
    >
      <div>
        <p
          className="thread-text"
        >
          2 comments
        </p>
        <div
          className="comment"
        >
          <p
            className="comment-title"
          >
            Jeff - 
          </p>
          <p
            className="comment-text"
          >
            This article seems to lack source material for the conclusions drawn in the discussion.
          </p>
        </div>
        <div
          className="comment"
        >
          <p
            className="comment-title"
          >
            mark - 
          </p>
          <p
            className="comment-text"
          >
            Does anyone else find figure 5 really confusing?
          </p>
        </div>
      </div>
      <div
        className="collapse"
        style={Object {}}
      >
        <div
          style={
            Object {
              "marginTop": "25px",
            }
          }
        >
          <p
            style={
              Object {
                "color": "black",
                "margin": "5px 20px",
              }
            }
          >
            Have something to say about this thread? Add a comment to join the discussion.
          </p>
          <textarea
            className="form-control"
            onChange={[Function]}
            placeholder="Your comment here"
            style={
              Object {
                "borderRadius": "6px",
                "margin": "15px auto",
                "padding": "8px",
                "width": "95%",
              }
            }
          />
          <div
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "flex-end",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <button
              aria-label={null}
              className="add article-button btn btn-secondary btn-sm"
              onClick={[Function]}
              type="button"
            >
              post comment
            </button>
            <button
              aria-label={null}
              className="back article-button btn btn-secondary btn-sm"
              onClick={[Function]}
              type="button"
            >
              cancel
            </button>
          </div>
        </div>
      </div>
      <div
        className="fade"
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
      >
        <button
          aria-label={null}
          className="view article-button btn btn-secondary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          comment
        </button>
      </div>
    </span>
  </div>
</div>
`;
